zerver/lib/narrow.py

```diff
@@ -861,6 +869,74 @@ def get_channel_from_narrow_access_unchecked(
     return None
 
 
+# This function corrects the conversation filters of the narrow - channel,
+# topic, DM conversation to point to the narrow in which the message
+# id operand of the `with` operator is present in.
+#
+# This is done as an effort to facilitate permalinks for topics -- if there
+# exists a message represented by the id (`with` operand), we redirect to
+# the narrow in which the message is present. Else, we fallback
+# to the narrow mentioned originally.
+def update_narrow_terms_containing_with_operator(
+    realm: Realm,
+    maybe_user_profile: Union[UserProfile, AnonymousUser],
+    narrow: Optional[List[NarrowParameter]],
+) -> Optional[List[NarrowParameter]]:
+    if narrow is None:
+        return narrow
+
+    contains_with_operator = list(filter(lambda term: term.operator == "with", narrow))
+
+    if len(contains_with_operator) > 1:
+        narrow = [term for term in narrow if term.operator != "with"]
+        return narrow
+
+    if contains_with_operator:
+        with_term = contains_with_operator[0]
+        narrow.remove(with_term)
+        try:
+            message_id = int(with_term.operand)
+        except ValueError:
+            return narrow
+
+        if maybe_user_profile.is_authenticated:
+            try:
+                message = access_message(maybe_user_profile, message_id)
+            except JsonableError:
+                return narrow
+        else:
+            try:
+                message = access_web_public_message(realm, message_id)
+            except MissingAuthenticationError:
+                return narrow
+
+        narrow = [
+            term
+            for term in narrow
+            if term.operator not in ["stream", "channel", "topic", "dm", "dm-including"]
+        ]
+
+        if message.recipient.type == Recipient.STREAM:
+            channel_id = message.recipient.type_id
+            topic = message.topic_name()
+            channel_topic_narrow = [
+                NarrowParameter(operator="channel", operand=channel_id),
+                NarrowParameter(operator="topic", operand=topic),
+            ]
+            narrow = channel_topic_narrow + narrow
+
+        elif message.recipient.type == Recipient.PERSONAL:
+            dm_narrow = [NarrowParameter(operator="dm", operand=[message.recipient.type_id])]
+            narrow = dm_narrow + narrow
+
+        elif message.recipient.type == Recipient.DIRECT_MESSAGE_GROUP:
+            huddle_user_ids = list(get_huddle_user_ids(message.recipient))
+            dm_narrow = [NarrowParameter(operator="dm", operand=huddle_user_ids)]
+            narrow = dm_narrow + narrow
+
+    return narrow
+
+
```

Comment:
This function's logic did not seem to match the specification. I've rewritten it to, I believe, do so. Haven't had a chance to update tests.

``` diff
diff --git a/zerver/lib/narrow.py b/zerver/lib/narrow.py
index ad3a3c10af..5ba90ed46e 100644
--- a/zerver/lib/narrow.py
+++ b/zerver/lib/narrow.py
@@ -208,6 +208,19 @@ class BadNarrowOperatorError(JsonableError):
         return _("Invalid narrow operator: {desc}")
 
 
+class InvalidOperatorCombinationError(JsonableError):
+    code = ErrorCode.BAD_NARROW
+    data_fields = ["desc"]
+
+    def __init__(self, desc: str) -> None:
+        self.desc: str = desc
+
+    @staticmethod
+    @override
+    def msg_format() -> str:
+        return _("Invalid narrow operator: {desc}")
+
+
 ConditionTransform: TypeAlias = Callable[[ClauseElement], ClauseElement]
 
 # These delimiters will not appear in rendered messages or HTML-escaped topics.
@@ -869,14 +882,17 @@ def get_channel_from_narrow_access_unchecked(
     return None
 
 
-# This function corrects the conversation filters of the narrow - channel,
-# topic, DM conversation to point to the narrow in which the message
-# id operand of the `with` operator is present in.
+# This function implements the core logic of the `with` operator,
+# which is designed to support permanent links to a topic that
+# robustly function if the topic is moved.
+#
+# The with operator accepts a message ID as an operand. If the
+# message ID does not exist or is otherwise not accessible to the
+# current user, then it has no effect.
 #
-# This is done as an effort to facilitate permalinks for topics -- if there
-# exists a message represented by the id (`with` operand), we redirect to
-# the narrow in which the message is present. Else, we fallback
-# to the narrow mentioned originally.
+# Otherwise, the narrow terms are mutated to remove any
+# channel/topic/dm operators, replacing them with the appropriate
+# operators for the conversation view containing the targeted message.
 def update_narrow_terms_containing_with_operator(
     realm: Realm,
     maybe_user_profile: Union[UserProfile, AnonymousUser],
@@ -885,56 +901,61 @@ def update_narrow_terms_containing_with_operator(
     if narrow is None:
         return narrow
 
-    contains_with_operator = list(filter(lambda term: term.operator == "with", narrow))
+    with_operator_terms = list(filter(lambda term: term.operator == "with", narrow))
 
-    if len(contains_with_operator) > 1:
+    if len(with_operator_terms) > 1:
+        raise InvalidOperatorCombinationError(_("Duplicate 'with' operators."))
         narrow = [term for term in narrow if term.operator != "with"]
         return narrow
+    elif len(with_operator_terms) == 0:
+        return narrow
+
+    with_term = with_operator_terms[0]
+    narrow.remove(with_term)
+    try:
+        message_id = int(with_term.operand)
+    except ValueError:
+        # TODO: This probably should be handled earlier.
+        raise BadNarrowOperatorError(_("Invalid 'with' operator"))
 
-    if contains_with_operator:
-        with_term = contains_with_operator[0]
-        narrow.remove(with_term)
+    if maybe_user_profile.is_authenticated:
+        try:
+            message = access_message(maybe_user_profile, message_id)
+        except JsonableError:
+            return narrow
+    else:
         try:
-            message_id = int(with_term.operand)
-        except ValueError:
+            message = access_web_public_message(realm, message_id)
+        except MissingAuthenticationError:
             return narrow
 
-        if maybe_user_profile.is_authenticated:
-            try:
-                message = access_message(maybe_user_profile, message_id)
-            except JsonableError:
-                return narrow
-        else:
-            try:
-                message = access_web_public_message(realm, message_id)
-            except MissingAuthenticationError:
-                return narrow
-
-        narrow = [
-            term
-            for term in narrow
-            if term.operator not in ["stream", "channel", "topic", "dm", "dm-including"]
-        ]
+    # TODO: Do we need the legacy names here, or can they have been
+    # already canonicalized in building a NarrowParameter?
+    filtered_terms = [
+        term
+        for term in narrow
+        if term.operator not in ["stream", "channel", "topic", "dm", "pm-with"]
+    ]
 
-        if message.recipient.type == Recipient.STREAM:
-            channel_id = message.recipient.type_id
-            topic = message.topic_name()
-            channel_topic_narrow = [
-                NarrowParameter(operator="channel", operand=channel_id),
-                NarrowParameter(operator="topic", operand=topic),
-            ]
-            narrow = channel_topic_narrow + narrow
+    if message.recipient.type == Recipient.STREAM:
+        channel_id = message.recipient.type_id
+        topic = message.topic_name()
+        channel_conversation_terms = [
+            NarrowParameter(operator="channel", operand=channel_id),
+            NarrowParameter(operator="topic", operand=topic),
+        ]
+        return channel_conversation_terms + filtered_terms
 
-        elif message.recipient.type == Recipient.PERSONAL:
-            dm_narrow = [NarrowParameter(operator="dm", operand=[message.recipient.type_id])]
-            narrow = dm_narrow + narrow
+    elif message.recipient.type == Recipient.PERSONAL:
+        dm_conversation_terms = [NarrowParameter(operator="dm", operand=[message.recipient.type_id])]
+        return dm_conversation_terms + filtered_terms
 
-        elif message.recipient.type == Recipient.DIRECT_MESSAGE_GROUP:
-            huddle_user_ids = list(get_huddle_user_ids(message.recipient))
-            dm_narrow = [NarrowParameter(operator="dm", operand=huddle_user_ids)]
-            narrow = dm_narrow + narrow
+    elif message.recipient.type == Recipient.DIRECT_MESSAGE_GROUP:
+        huddle_user_ids = list(get_huddle_user_ids(message.recipient))
+        dm_conversation_terms = [NarrowParameter(operator="dm", operand=huddle_user_ids)]
+        return dm_conversation_terms + filtered_terms
 
-    return narrow
+    raise AssertionError("Invalid recipient type")
 
 
 def exclude_muting_conditions(
```